### ISSUE TITLE
Error in SaveChanges() when EntityTypeMappings has multiple Fragments

### ISSUE THREAD

**Issue by lbouza** [thepirat000/Audit.NET#409]
Hi,
We are having an issue related to #273 
We have an inheritance of four levels being `RootClass` and `MiddleClass` abstract classes. The problem occurs when we try to add an instance of `LeafTwoClass`.

**DataBase Model**
```
    public class BaseEntity
    {
        public int Id { get; set; }

        //more properties
    }
    public abstract class RootClass:BaseEntity
    {
        public string PropertyOne { get; set; }
        public string PropertyTwo { get; set; }
    }
    public abstract class MiddleClass:RootClass
    {
        public string PropertyThree { get; set; }
    }
    public class LeafOneClass:RootClass
    {
        public string PropertyFour { get; set; }
    }
    public class LeafTwoClass:MiddleClass
    {
        public string PropertyFive { get; set; }
    }
```

**Fluent API Mapping**
```
    public class BaseEntityMap<T> : EntityTypeConfiguration<T>        where T : BaseEntity
    {
        public BaseEntityMap()
        {
            HasKey(t => t.Id);
        }
    }
    public class RootClassMap:BaseEntityMap<RootClass>
    {
        public RootClassMap()
        {
            ToTable("RootClasses");
        }
    }
    public class LeafOneClassMap:BaseEntityMap<LeafOneClass>
    {
        public LeafOneClassMap()
        {
            ToTable("LeafOneClasses");
        }
    }
    public class LeafTwoClassMap:BaseEntityMap<LeafTwoClass>
    {
        public LeafTwoClassMap()
        {
            ToTable("LeafTwoClasses");
        }
    }
```

**DbContext**
```
    public class DataBaseContext:DbContext
    {
        private static DbContextHelper _helper = new DbContextHelper();
        private readonly IAuditDbContext _auditContext;

        public DataBaseContext() : base("DataBaseAuditTests")
        {
            _auditContext = new DefaultAuditContext(this);
            _helper.SetConfig(_auditContext);

            Audit.EntityFramework.Configuration.Setup().ForContext<DataBaseContext>(config =>
            {
                config.IncludeEntityObjects();
                config.IncludeIndependantAssociations();
            }).UseOptOut();
        }

        public override int SaveChanges()
        {
            return _helper.SaveChanges(_auditContext, () => base.SaveChanges());
        }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.Configurations.Add(new RootClassMap());
            modelBuilder.Configurations.Add(new LeafOneClassMap());
            modelBuilder.Configurations.Add(new LeafTwoClassMap());
        }
    }
```

**Exception**
```
Sequence contains more than one element
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at EntityKeyHelper.GetMappingFragment(Type type, DbContext context)
   at EntityKeyHelper.GetTableName(Type type, DbContext context)
   at Audit.EntityFramework.DbContextHelper.GetEntityName(DbContext dbContext, Object entity)
   at Audit.EntityFramework.DbContextHelper.CreateAuditEvent(IAuditDbContext context)
   at Audit.EntityFramework.DbContextHelper.SaveChanges(IAuditDbContext context, Func`1 baseSaveChanges)
  ...
```

To reproduce the issue we attach a project sample
https://github.com/lbouza/AuditTest/tree/master

We are using the next libraries:
 `Audit.Net 17.0.4`
`Audit.EntityFramework 17.0.4`
`EntityFramework 6.2`

**Comment by thepirat000** [comment]
I've tested with the repository you provided but I was unable to reproduce, it's not throwing the exception.

![image](https://user-images.githubusercontent.com/9836380/120569003-ac8ada80-c3da-11eb-9065-8d4868e652da.png)

**Comment by lbouza** [comment]
Thanks for the quick response.
What version of EntityFramework package did you add

**Comment by thepirat000** [comment]
6.2

**Comment by thepirat000** [comment]
I was able to reproduce. Will take a look and let you know

**Comment by thepirat000** [comment]
The only way I can think to fix this, is to change the last `Single()` to `First()`, on this line:

https://github.com/thepirat000/Audit.NET/blob/532d3fd8b48750a12cdba8f2420a733d58a32e54/src/Audit.EntityFramework/EntityKeyHelper.cs#L119

That will make the library depend on the internal order that EF gives to the mapping Fragments, so assuming the leaf classes are always at the beginning of the fragments list.

I've made some tests and it seems to work fine

**Comment by thepirat000** [comment]
This was fixed on version 17.0.6. 

All tests are passing, please upgrade your references and re-test

**Comment by lbouza** [comment]
The feature is working as expected. Thanks for your help and thanks for the library.
